### PR TITLE
Ensure resumed tasks are not accidentally forgotten

### PR DESCRIPTION
**Pull request by fjetter**
closes https://github.com/dask/distributed/issues/6194
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

Three changes in this PR

- Fix an erroneous transition resumed->released which previously actually released a key but instead it should transition to cancelled if the task is still being processed.
- Removed a redundant cancelled->resumed transition which makes the logs **much** less verbose and easier to read. This transition was only an indirection and I instead _inlined_ the code at the two places where this matters. The transition log now reads as expected in these situations
- Ensure that `ts._next` is not set for cancelled tasks. Cancelled tasks should *always* transition to released once they are done.

### CODE REVIEW

File: distributed/utils_test.py [hunk 1]
```diff
@@ -87,7 +87,7 @@
     if isinstance(logger, logging.Logger)
 }
 
-_TEST_TIMEOUT = 30
+_TEST_TIMEOUT = 30 if not sys.gettrace() else 3600
```
Comment (fjetter) [inline distributed/utils_test.py:90]: TODO
- [ ] factor this out to another PR

Comment (fjetter) [reply distributed/utils_test.py:90]: https://github.com/dask/distributed/pull/6218

**Comment by fjetter** [comment]
The windows test failure is a known issue

https://github.com/dask/distributed/issues/6147

```
File "D:\a\distributed\distributed\distributed\worker.py", line 4053, in validate_task

    self.validate_task_fetch(ts)

  File "D:\a\distributed\distributed\distributed\worker.py", line 3995, in validate_task_fetch

    assert ts.who_has

AssertionError

2022-04-27 11:54:02,400 - distributed.core - ERROR - Invalid TaskState encountered for <TaskState "('arange-sum-1960f5e08fa9c9689f5e9ef3be470[377](https://github.com/dask/distributed/runs/6192747991?check_suite_focus=true#step:11:377)', 8)" fetch>.

Story:

[("('arange-sum-1960f5e08fa9c9689f5e9ef3be470377', 8)", 'compute-task', 'compute-task-1651060440.6118023', 1651060440.640618), ("('arange-sum-1960f5e08fa9c9689f5e9ef3be470377', 8)", 'released', 'waiting', 'waiting', {"('arange-sum-1960f5e08fa9c9689f5e9ef3be470377', 8)": 'ready'}, 'compute-task-1651060440.6118023', 1651060440.6406472), ("('arange-sum-1960f5e08fa9c9689f5e9ef3be470377', 8)", 'waiting', 'ready', 'ready', {}, 'compute-task-1651060440.6118023', 1651060440.6406634), ("('arange-sum-1960f5e08fa9c9689f5e9ef3be470377', 8)", 'ready', 'executing', 'executing', {}, 'task-finished-1651060440.648988', 1651060440.6492662), ("('arange-sum-1960f5e08fa9c9689f5e9ef3be470377', 8)", 'put-in-memory', 'task-finished-1651060440.6506114', 1651060440.6508203), ("('arange-sum-1960f5e08fa9c9689f5e9ef3be470377', 8)", 'executing', 'memory', 'memory', {"('arange-sum-59ee542fb6a8f7c418bd1ceadfe68d11', 8)": 'executing'}, 'task-finished-1651060440.6506114', 1651060440.65087), ("('arange-sum-1960f5e08fa9c9689f5e9ef3be470377', 8)", 'ensure-task-exists', 'memory', 'compute-task-1651060440.6795359', 1651060440.705263), ('free-keys', ("('arange-sum-1960f5e08fa9c9689f5e9ef3be470377', 8)",), 'task-finished-1651060440.7063837', 1651060440.743776), ("('arange-sum-1960f5e08fa9c9689f5e9ef3be470377', 8)", 'release-key', 'task-finished-1651060440.7063837', 1651060440.74[379](https://github.com/dask/distributed/runs/6192747991?check_suite_focus=true#step:11:379)23), ("('arange-sum-1960f5e08fa9c9689f5e9ef3be470377', 8)", 'memory', 'released', 'released', {}, 'task-finished-1651060440.706[383](https://github.com/dask/distributed/runs/6192747991?check_suite_focus=true#step:11:383)7', 1651060440.74[384](https://github.com/dask/distributed/runs/6192747991?check_suite_focus=true#step:11:384)28), ("('arange-sum-1960f5e08fa9c9689f5e9ef3be470377', 8)", 'released', 'forgotten', 'forgotten', {}, 'client-releases-keys-1651060442.0950112', 1651060442.1340976), ("('arange-sum-1960f5e08fa9c9689f5e9ef3be470377', 8)", 'compute-task', 'compute-task-1651060442.1040328', 1651060442.1347113), ("('arange-sum-1960f5e08fa9c9689f5e9ef3be470377', 8)", 'released', 'waiting', 'waiting', {"('arange-sum-1960f5e08fa9c9689f5e9ef3be470377', 8)": 'ready'}, 'compute-task-1651060442.1040328', 1651060442.1347415), ("('arange-sum-1960f5e08fa9c9689f5e9ef3be470377', 8)", 'waiting', 'ready', 'ready', {}, 'compute-task-1651060442.1040328', 1651060442.1347587), ("('arange-sum-1960f5e08fa9c9689f5e9ef3be470377', 8)", 'ready', 'executing', 'executing', {}, 'task-finished-1651060442.1677341', 1651060442.1682491), ("('arange-sum-1960f5e08fa9c9689f5e9ef3be470377', 8)", 'put-in-memory', 'task-finished-1651060442.169614', 1651060442.1697793), ("('arange-sum-1960f5e08fa9c9689f5e9ef3be470377', 8)", 'executing', 'memory', 'memory', {"('arange-sum-b267f04bd650ea06d6abf5234529c789', 8)": 'executing'}, 'task-finished-1651060442.169614', 1651060442.1698287), ('free-keys', ("('arange-sum-1960f5e08fa9c9689f5e9ef3be470377', 8)",), 'client-releases-keys-1651060442.290155', 1651060442.3110044), ("('arange-sum-1960f5e08fa9c9689f5e9ef3be470377', 8)", 'release-key', 'client-releases-keys-1651060442.290155', 1651060442.3110235), ("('arange-sum-1960f5e08fa9c9689f5e9ef3be470377', 8)", 'memory', 'released', 'released', {"('arange-sum-1960f5e08fa9c9689f5e9ef3be470377', 8)": 'forgotten'}, 'client-releases-keys-1651060442.290155', 1651060442.311108), ("('arange-sum-1960f5e08fa9c9689f5e9ef3be470377', 8)", 'released', 'forgotten', 'forgotten', {}, 'client-releases-keys-1651060442.290155', 1651060442.3111255), ("('arange-sum-1960f5e08fa9c9689f5e9ef3be470377', 8)", 'compute-task', 'compute-task-1651060442.3011189', 1651060442.3126717), ("('arange-sum-1960f5e08fa9c9689f5e9ef3be470377', 8)", 'released', 'waiting', 'waiting', {"('arange-sum-1960f5e08fa9c9689f5e9ef3be470377', 8)": 'ready'}, 'compute-task-1651060442.3011189', 1651060442.3127015), ("('arange-sum-1960f5e08fa9c9689f5e9ef3be470377', 8)", 'waiting', 'ready', 'ready', {}, 'compute-task-1651060442.3011189', 1651060442.3127184), ("('arange-sum-1960f5e08fa9c9689f5e9ef3be470377', 8)", 'ensure-task-exists', 'ready', 'compute-task-1651060442.3024514', 1651060442.3128352), ("('arange-sum-1960f5e08fa9c9689f5e9ef3be470377', 8)", 'release-key', 'compute-task-1651060442.3024514', 1651060442.3128846), ("('arange-sum-1960f5e08fa9c9689f5e9ef3be470377', 8)", 'ready', 'released', 'released', {}, 'compute-task-1651060442.3024514', 1651060442.3129065), ("('arange-sum-1960f5e08fa9c9689f5e9ef3be470377', 8)", 'released', 'fetch', 'fetch', {}, 'compute-task-1651060442.3024514', 1651060442.312925), ("('arange-sum-1960f5e08fa9c9689f5e9ef3be470377', 8)", 'ready', 'fetch', 'fetch', {}, 'compute-task-1651060442.3024514', 1651060442.31293)]
```

**Comment by mrocklin** [comment]
I took a brief look at this.  No objection from me, but I didn't dive deeply into the logic.  If tests pass and you're feeling confident about the added value @fjetter I think that it's ok to merge.  If you can get someone like @crusaderky or @gjoseph92 to take a look that would be better of course.

File: distributed/worker.py [hunk 2]
```diff
@@ -2149,8 +2148,7 @@ def transition_cancelled_error(
                 traceback_text,
                 stimulus_id=stimulus_id,
             )
-        if ts._next:
-            recs[ts] = ts._next
+        recs[ts] = "released"
```
Comment (gjoseph92) [inline distributed/worker.py:2151]: Why override what the transition function we just called recommended to do?

File: distributed/worker.py [hunk 3]
```diff
@@ -2231,18 +2229,29 @@ def _transition_from_resumed(
...
+            # If we're not done, yet, just remember where we want to be next
+            ts._next = finish
+        else:
+            # Flight/executing finished unsuccesfully, i.e. not in memory
+            assert finish != "memory"
             next_state = ts._next
-            # if the next state is already intended to be waiting or if the
-            # coro/thread is still running (ts.done==False), this is a noop
+            assert next_state in {"waiting", "fetch"}
+            assert ts._previous in {"executing", "flight"}
```
Comment (gjoseph92) [inline distributed/worker.py:2247]:
```suggestion
            assert next_state in {"waiting", "fetch"}, next_state
            assert ts._previous in {"executing", "flight"}, ts._previous
```

File: distributed/worker.py [hunk 4]
```diff
@@ -2261,6 +2270,14 @@ def transition_resumed_missing(
         """
         return self._transition_from_resumed(ts, "missing", stimulus_id=stimulus_id)
 
+    def transition_resumed_released(self, ts: TaskState, *, stimulus_id: str):
+        if not ts.done:
+            ts.state = "cancelled"
+            ts._next = None
+            return {}, []
```
Comment (gjoseph92) [inline distributed/worker.py:2277]: What's going to eventually pick this up and move it out of cancelled if there are no recommendations and no next?

**Review by gjoseph92 — COMMENTED** [review]
I've never understood the flow for cancelled and released tasks, so I don't think I'm in a good position to review this. I find these states incredibly confusing. I'm looking forward to this hopefully making more sense with https://github.com/dask/distributed/issues/5895.

File: distributed/worker.py (hunk 2 above)
Comment (fjetter) [reply distributed/worker.py:2151]: Good catch. This entire transition is a bit off for "historic" reasons (i.e. I got confused during the initial refactor). The cancellation->error transition can only be reached from an executing state, i.e. previous is always executing and the "flight" path is dead code.
Further, this error transition will *never* generate a recommendation for this key since error is supposed to be dead end unless the scheduler releases the key. However, in this specific situation, the scheduler _already_ released the key which is why we'll need to recommend the release of this key.

I'll updated this transition accordingly and added two tests for it

File: distributed/worker.py (hunk 4 above)
Comment (fjetter) [reply distributed/worker.py:2273]: TLDR Once `ts.done = True` is set, i.e. execute/flight is done


`ts._next` should never have been set for `cancelled`. When I implemented cancelled/resumed states I made a few mistakes. The only relevant _next_ state for cancelled is `released`. That's the entire point of the state. The worker was instructed to release a key but it can't because it is "stuck" waiting for _something_ to finish, i.e. either the execution thread or the gather_data coroutine.
Once execution/gather finishes, they'll recommend a transition, e.g. upon success they'll recommend a transition to memory. For example cancelled->memory will ensure the key is released.

Why is this logic not directly implemented as part of the gather_dep/execute result parser? Well, have a look at the code there. Particularly the gather_dep result parser/finally clause is the most frequent source of deadlocks because the logic just blows up.
There is a bit of a design philosophy behind this to break a big, complex decision up into many small decisions that can be made using _local context_ information.

Consider the following example

- T1 was instructed to be computed
- T1 is dispatched to the threadpool
- T1 is requested to be released
- T1 finishes

The result, i.e. once it finishes could be implemented as

```python
if result == "success":
    if ts.not_cancelled:
        put_key_in_memory()
    else:
        release_key()
else:
    result == "failed"
    
    if ts.not_cancelled:
        if ts.asked_to_be_fetched_instead:
            # Whether this is a valid thing for the scheduler to ask is out
            # of scope for this comment. It happens/happened
            reschedule_to_fetch_key()
        else:
            put_key_in_memory()
    else:
        release_key()
```

With this transition system, it instead becomes

```python
# executing result parser
# This only requires local context, the decision should be simple, straight forward
if result == "success":
    recommend_memory()
else:
    assert result == "success"
    recommend_error()

def transition_cancelled_error(...):
    assert stuff
    release_task()

def transition_cancelled_memory(...):
    assert stuff
    put_key_in_memory()
```

This decision tree is a bit more complex for gathering keys. I'm not 100% convinced anymore if this is the right approach but here we are right now. The recent refactoring will allow us mid-term to move away from this if we choose to do so.

File: distributed/tests/test_cancelled_state.py [hunk 5]
```diff
@@ -240,3 +235,133 @@ async def get_data(self, comm, *args, **kwargs):
...
+    lock_executing = Lock()
+
+    def block_execution(x, event, lock):
+        event.set()
+        with lock:
+            return x
+
+    class BlockedGetData(Worker):
+        async def get_data(self, comm, *args, **kwargs):
+            in_get_data.set()
```
Comment (crusaderky) [inline distributed/tests/test_cancelled_state.py:255]:
```suggestion
            await in_get_data.set()
```

File: distributed/tests/test_cancelled_state.py [hunk 6]
```diff
@@ -240,3 +235,133 @@ async def get_data(self, comm, *args, **kwargs):
...
         # Everything should still be executing as usual after this
         assert await c.submit(sum, c.map(inc, range(10))) == sum(map(inc, range(10)))
+
+
+@gen_cluster(client=True, nthreads=[("", 1)])
+async def test_in_flight_lost_after_resumed(c, s, b):
+    block_get_data = asyncio.Lock()
+    in_get_data = asyncio.Event()
+
+    executing = Event()
```
Comment (crusaderky) [inline distributed/tests/test_cancelled_state.py:245]: This event is set by block_execution, but there's nothing waiting for it

File: distributed/tests/test_cancelled_state.py [hunk 7]
```diff
@@ -240,3 +235,133 @@ async def get_data(self, comm, *args, **kwargs):
...
+        fut1 = c.submit(
+            block_execution,
+            1,
+            executing,
+            lock_executing,
+            workers=[a.address],
+            allow_other_workers=True,
+        )
+        await fut1
+        await lock_executing.acquire()
```
Comment (crusaderky) [inline distributed/tests/test_cancelled_state.py:269]: This looks like a very complicated way of testing that fut1 is not recomputed?

File: distributed/tests/test_cancelled_state.py [hunk 8]
```diff
@@ -240,3 +235,133 @@ async def get_data(self, comm, *args, **kwargs):
...
+        in_get_data.clear()
```
Comment (crusaderky) [inline distributed/tests/test_cancelled_state.py:270]:
```suggestion
        await in_get_data.clear()
```

File: distributed/tests/test_cancelled_state.py [hunk 9]
```diff
@@ -240,3 +235,133 @@ async def get_data(self, comm, *args, **kwargs):
...
+        await block_get_data.acquire()
+        fut2 = c.submit(inc, fut1, workers=[b.address])
+
+        await in_get_data.wait()
+        await s.remove_worker(a.address, "foo", close=False, safe=True)
+
+        while not b.tasks[fut1.key].state == "resumed":
+            await asyncio.sleep(0.1)
```
Comment (crusaderky) [inline distributed/tests/test_cancelled_state.py:278]: Am I missing a reason why fut1 is guaranteed to already b in b here?
```suggestion
        assert fut1.key in b.tasks
        while b.tasks[fut1.key].state != "resumed":
            await asyncio.sleep(0.01)
```

File: distributed/tests/test_cancelled_state.py [hunk 10]
```diff
@@ -240,3 +235,133 @@ async def get_data(self, comm, *args, **kwargs):
...
+
+        fut1.release()
+        fut2.release()
+        await asyncio.sleep(0.2)
```
Comment (crusaderky) [inline distributed/tests/test_cancelled_state.py:282]: What's the purpose of this?

File: distributed/tests/test_cancelled_state.py [hunk 11]
```diff
@@ -240,3 +235,133 @@ async def get_data(self, comm, *args, **kwargs):
...
+        block_get_data.release()
+        while b.tasks:
+            await asyncio.sleep(0.1)
```
Comment (crusaderky) [inline distributed/tests/test_cancelled_state.py:285]:
```suggestion
            await asyncio.sleep(0.01)
```

File: distributed/tests/test_cancelled_state.py [hunk 12]
```diff
@@ -240,3 +235,133 @@ async def get_data(self, comm, *args, **kwargs):
...
+            in_get_data.set()
+            async with block_get_data:
+                return await super().get_data(comm, *args, **kwargs)
+
+    async with BlockedGetData(s.address, name="blocked-get-dataworker") as a:
+        fut1 = c.submit(
+            block_execution,
+            1,
+            executing,
+            lock_executing,
```
Comment (crusaderky) [inline distributed/tests/test_cancelled_state.py:264]:
```suggestion
            lock_executing,
            key="fut1",
```

File: distributed/tests/test_cancelled_state.py [hunk 13]
```diff
@@ -240,3 +235,133 @@ async def get_data(self, comm, *args, **kwargs):
...
+            workers=[a.address],
+            allow_other_workers=True,
+        )
+        await fut1
+        await lock_executing.acquire()
+        in_get_data.clear()
+        await block_get_data.acquire()
+        fut2 = c.submit(inc, fut1, workers=[b.address])
```
Comment (crusaderky) [inline distributed/tests/test_cancelled_state.py:272]:
```suggestion
        fut2 = c.submit(inc, fut1, key="fut2", workers=[b.address])
```

File: distributed/tests/test_cancelled_state.py [hunk 14]
```diff
@@ -240,3 +235,133 @@ async def get_data(self, comm, *args, **kwargs):
...
+    )
+
+
+@gen_cluster(client=True)
+async def test_cancelled_error(c, s, a, b):
+    executing = Event()
+    lock_executing = Lock()
+    await lock_executing.acquire()
+
+    def block_execution(x, event, lock):
```
Comment (crusaderky) [inline distributed/tests/test_cancelled_state.py:305]:
```suggestion
    def block_execution(event, lock):
```

File: distributed/tests/test_cancelled_state.py [hunk 15]
```diff
@@ -240,3 +235,133 @@ async def get_data(self, comm, *args, **kwargs):
...
+        event.set()
+        with lock:
+            raise RuntimeError()
+
+    fut1 = c.submit(
+        block_execution,
+        1,
+        executing,
+        lock_executing,
```
Comment (crusaderky) [inline distributed/tests/test_cancelled_state.py:314]:
```suggestion
        executing,
        lock_executing,
        key="fut1",
```

File: distributed/tests/test_cancelled_state.py [hunk 16]
```diff
@@ -240,3 +235,133 @@ async def get_data(self, comm, *args, **kwargs):
...
+        workers=[b.address],
+        allow_other_workers=True,
+    )
+
+    await executing.wait()
+    assert b.tasks[fut1.key].state == "executing"
+    fut1.release()
+    while b.tasks[fut1.key].state == "executing":
+        await asyncio.sleep(0.1)
```
Comment (crusaderky) [inline distributed/tests/test_cancelled_state.py:323]:
```suggestion
        await asyncio.sleep(0.01)
```

File: distributed/tests/test_cancelled_state.py [hunk 17]
```diff
@@ -240,3 +235,133 @@ async def get_data(self, comm, *args, **kwargs):
...
+    await lock_executing.release()
+    while b.tasks:
+        await asyncio.sleep(0.1)
```
Comment (crusaderky) [inline distributed/tests/test_cancelled_state.py:326]:
```suggestion
        await asyncio.sleep(0.01)
```

File: distributed/tests/test_cancelled_state.py [hunk 18]
```diff
@@ -240,3 +235,133 @@ async def get_data(self, comm, *args, **kwargs):
...
+    )
+
+
+@gen_cluster(client=True, nthreads=[("", 1, {"resources": {"A": 1}})])
+async def test_cancelled_error_with_ressources(c, s, a):
+    executing = Event()
+    lock_executing = Lock()
+    await lock_executing.acquire()
+
+    def block_execution(x, event, lock):
```
Comment (crusaderky) [inline distributed/tests/test_cancelled_state.py:343]:
```suggestion
    def block_execution(event, lock):
```

File: distributed/tests/test_cancelled_state.py [hunk 19]
```diff
@@ -240,3 +235,133 @@ async def get_data(self, comm, *args, **kwargs):
...
+        event.set()
+        with lock:
+            raise RuntimeError()
+
+    fut1 = c.submit(
+        block_execution,
+        1,
+        executing,
+        lock_executing,
```
Comment (crusaderky) [inline distributed/tests/test_cancelled_state.py:352]:
```suggestion
        lock_executing,
        key="fut1",
```

File: distributed/tests/test_cancelled_state.py [hunk 20]
```diff
@@ -240,3 +235,133 @@ async def get_data(self, comm, *args, **kwargs):
...
+        resources={"A": 1},
+    )
+
+    await executing.wait()
+    fut2 = c.submit(inc, 1, resources={"A": 1})
```
Comment (crusaderky) [inline distributed/tests/test_cancelled_state.py:357]:
```suggestion
    fut2 = c.submit(inc, 1, key="fut2", resources={"A": 1})
```

File: distributed/tests/test_cancelled_state.py (hunk 18 above)
Comment (crusaderky) [inline distributed/tests/test_cancelled_state.py:343]:
```suggestion
    def block_execution(event, lock):
```

File: distributed/tests/test_cancelled_state.py [hunk 21]
```diff
@@ -240,3 +235,133 @@ async def get_data(self, comm, *args, **kwargs):
...
+    await lock_executing.acquire()
+
+    def block_execution(x, event, lock):
+        event.set()
+        with lock:
+            raise RuntimeError()
+
+    fut1 = c.submit(
+        block_execution,
+        1,
```
Comment (crusaderky) [inline distributed/tests/test_cancelled_state.py:350]:
```suggestion
```

File: distributed/tests/test_cancelled_state.py [hunk 22]
```diff
@@ -240,3 +235,133 @@ async def get_data(self, comm, *args, **kwargs):
...
+
+    await executing.wait()
+    fut2 = c.submit(inc, 1, resources={"A": 1})
+
+    while fut2.key not in a.tasks:
+        await asyncio.sleep(0.1)
+
+    fut1.release()
+    while a.tasks[fut1.key].state == "executing":
+        await asyncio.sleep(0.1)
```
Comment (crusaderky) [inline distributed/tests/test_cancelled_state.py:364]:
```suggestion
        await asyncio.sleep(0.01)
```
Shouldn't fut1 disappear from a entirely? Isn't there a race condition here?

File: distributed/worker.py [hunk 23]
```diff
@@ -2129,29 +2128,24 @@ def transition_cancelled_error(
...
-                exception,
-                traceback,
-                exception_text,
-                traceback_text,
-                stimulus_id=stimulus_id,
-            )
-        if ts._next:
-            recs[ts] = ts._next
-        return recs, instructions
+        assert ts._previous == "executing"
```
Comment (crusaderky) [inline distributed/worker.py:2131]: why not long-running?

File: distributed/worker.py [hunk 24]
```diff
@@ -2129,29 +2128,24 @@ def transition_cancelled_error(
...
+        # released this task
+        recs, _ = self.transition_executing_error(
+            ts,
+            exception,
+            traceback,
+            exception_text,
+            traceback_text,
+            stimulus_id=stimulus_id,
+        )
+        # Workers should never "retry" tasks. A to error transition should, by
```
Comment (crusaderky) [inline distributed/worker.py:2143]:
```suggestion
        # Workers should never "retry" tasks. A transition to error should, by
```

File: distributed/worker.py [hunk 25]
```diff
@@ -2231,18 +2225,29 @@ def _transition_from_resumed(
...
+        else:
+            # Flight/executing finished unsuccesfully, i.e. not in memory
+            assert finish != "memory"
             next_state = ts._next
-            # if the next state is already intended to be waiting or if the
-            # coro/thread is still running (ts.done==False), this is a noop
+            assert next_state in {"waiting", "fetch"}, next_state
+            assert ts._previous in {"executing", "flight"}, ts._previous
+
             if ts._next != finish:
```
Comment (crusaderky) [inline distributed/worker.py:2245]:
```suggestion
            if next_state != finish:
```

File: distributed/worker.py [hunk 26]
```diff
@@ -2261,6 +2266,14 @@ def transition_resumed_missing(
         """
         return self._transition_from_resumed(ts, "missing", stimulus_id=stimulus_id)
 
+    def transition_resumed_released(self, ts: TaskState, *, stimulus_id: str):
```
Comment (crusaderky) [inline distributed/worker.py:2269]:
```suggestion
    def transition_resumed_released(self, ts: TaskState, *, stimulus_id: str) -> RecsInsts:
```

File: distributed/tests/test_cancelled_state.py (hunk 10 above)
Comment (fjetter) [reply distributed/tests/test_cancelled_state.py:282]: replaced it by waiting for an explicit state (cancelled)

File: distributed/worker.py (hunk 23 above)
Comment (fjetter) [reply distributed/worker.py]: added it here but couldn't produce a test.

**Comment by fjetter** [comment]
I believe at least some of the test failures relate to https://github.com/dask/distributed/pull/5910

File: distributed/worker.py [hunk 27]
```diff
@@ -2129,29 +2128,24 @@ def transition_cancelled_error(
...
-                exception,
-                traceback,
-                exception_text,
-                traceback_text,
-                stimulus_id=stimulus_id,
-            )
-        if ts._next:
-            recs[ts] = ts._next
-        return recs, instructions
+        assert ts._previous == "executing" or ts.key in self.long_running
```
Comment (crusaderky) [inline distributed/worker.py:2131]: Isn't this equivalent?
```suggestion
        assert ts._previous in ("executing", "long-running")
```

Comment (fjetter) [reply distributed/worker.py:2209]: This has been used in other places of the code. I think it should be equivalent, I'll keep it like this for now

File: distributed/worker.py [hunk 28]
```diff
@@ -2207,29 +2206,24 @@ def transition_cancelled_error(
...
+        # We'll ignore instructions, i.e. we choose to not submit the failure
+        # message to the scheudler since from the schedulers POV it already
```
Comment (crusaderky) [inline distributed/worker.py:2211]:
```suggestion
        # message to the scheduler since from the schedulers POV it already
```

File: distributed/worker.py [hunk 29]
```diff
@@ -2207,29 +2206,24 @@ def transition_cancelled_error(
...
-                stimulus_id=stimulus_id,
-            )
-        if ts._next:
-            recs[ts] = ts._next
-        return recs, instructions
+        assert ts._previous == "executing" or ts.key in self.long_running
+        # We'll ignore instructions, i.e. we choose to not submit the failure
+        # message to the scheduler since from the schedulers POV it already
+        # released this task
+        recs, _ = self.transition_executing_error(
```
Comment (crusaderky) [inline distributed/worker.py:2213]: All transitions from executing call ensure_computing.
This should deadlock the worker if there are any tasks in ready state.

**Review by crusaderky — CHANGES_REQUESTED** [review]
transition_cancelled_error is accidentally dropping Execute instructions

File: distributed/worker.py (hunk 29 above)
Comment (fjetter) [reply distributed/worker.py:2213]: I added a test for this. This does not deadlock since the transition generates a recommendation. Only after acting on that recommendation we'll get an instruction.

Comment (crusaderky) [inline distributed/worker.py:2213]: could you assert this here?
```python
recs, instructions = self.transition_executing_error(
    ...
)
if self.validate:
    assert len(instructions) == 1
    assert isinstance(instructions[0], TaskErredMsg)
    assert instructions[0].key == ts.key
instructions.clear()
```

Comment (fjetter) [reply distributed/worker.py:2213]: sure

File: distributed/tests/test_cancelled_state.py [hunk 30]
```diff
@@ -198,6 +194,9 @@ async def wait_and_raise(*args, **kwargs):
 
     fut = c.submit(wait_and_raise)
     await wait_for_state(fut.key, "executing", w)
+    # Queue up another task to ensure this is not affected by our error handling
+    fut2 = c.submit(inc, 1)
+    await wait_for_state(fut2.key, "ready", w)
```
Comment (fjetter) [inline distributed/tests/test_cancelled_state.py:199]: @crusaderky this triggers the condition you are concerned about in the transition function about dropped instructions.

The task is queued up and we'll receive a recommendation. The only instruction at that point is the TaskErred message.

**Comment by crusaderky** [comment]
I think there may be a regression here: https://github.com/dask/distributed/issues/6305#issuecomment-1125953078
